tests/mutex.c

```diff
@@ -4,39 +4,82 @@
 #include <thread.h>
 #include <ktest.h>
 
-static mtx_t mtx = MTX_INITIALIZER(MTX_DEF);
-static volatile int32_t value;
+static mtx_t counter_mtx = MTX_INITIALIZER(MTX_DEF);
+static volatile int32_t counter_value;
 
-#define N 1000
-#define T 5
+#define COUNTER_N 1000
+#define COUNTER_T 5
 
-static thread_t *td[T];
+static thread_t *counter_td[COUNTER_T];
 
-static void mtx_test_main(void *arg) {
-  for (size_t i = 0; i < N; i++) {
-    mtx_lock(&mtx);
-    value++;
-    mtx_unlock(&mtx);
+static void counter_routine(void *arg) {
+  for (size_t i = 0; i < COUNTER_N; i++) {
+    mtx_lock(&counter_mtx);
+    int32_t v = counter_value;
+    thread_yield();
+    counter_value = v + 1;
+    mtx_unlock(&counter_mtx);
   }
 }
 
-static int mtx_test(void) {
-  value = 0;
+static int mtx_test_counter(void) {
+  counter_value = 0;
 
-  td[0] = thread_create("td1", mtx_test_main, NULL);
-  td[1] = thread_create("td2", mtx_test_main, NULL);
-  td[2] = thread_create("td3", mtx_test_main, NULL);
-  td[3] = thread_create("td4", mtx_test_main, NULL);
-  td[4] = thread_create("td5", mtx_test_main, NULL);
+  for (int i = 0; i < COUNTER_T; i++) {
+    char name[20];
+    snprintf(name, sizeof(name), "td%d", i);
+    counter_td[i] = thread_create(name, counter_routine, NULL);
+  }
+
+  for (int i = 0; i < COUNTER_T; i++)
+    sched_add(counter_td[i]);
+  for (int i = 0; i < COUNTER_T; i++)
+    thread_join(counter_td[i]);
+
+  assert(counter_value == COUNTER_N * COUNTER_T);
+
+  return KTEST_SUCCESS;
+}
+
+static mtx_t simple_mtx = MTX_INITIALIZER(MTX_DEF);
+static thread_t *simple_td0;
+static int simple_status = 0;
+
+static void simple_routine(void *arg) {
+  WITH_NO_PREEMPTION {
+    simple_status = 1;
+    mtx_lock(&simple_mtx);
+  }
+  WITH_NO_PREEMPTION {
+    simple_status = 2;
+    mtx_unlock(&simple_mtx);
+  }
+  simple_status = 3;
+}
+
+static int mtx_test_simple(void) {
+  simple_td0 = thread_create("td0", simple_routine, NULL);
+
+  mtx_lock(&simple_mtx);
+
+  sched_add(simple_td0);
+
+  while (simple_status != 1) {
+    thread_yield();
+  }
 
-  for (int i = 0; i < T; i++)
-    sched_add(td[i]);
-  for (int i = 0; i < T; i++)
-    thread_join(td[i]);
+  assert(simple_status == 1);
+  mtx_unlock(&simple_mtx);
 
-  assert(value == N * T);
+  thread_join(simple_td0);
+  assert(simple_status == 3);
+  // NOTE This assert is implementation-specific
+  // Some changes might make it invalid
+  // We could assert =!mtx_owned(...)= but that would be less restrictive
```

Comment:
Please use Doxygen style instead of Org-mode style here: `=!mtx_owned(...)=`.
And please use `/* ... */` comment instead of `// ...` to be more consistent with other Mimiker source files.

Reply:
I tried changing it appropriately. Doxygen supposedly supports Markdown formatting but I'm not sure if it worked out because I use neither of these.